### ISSUE TITLE
SwiftPM should support `swift install`

### ISSUE THREAD

**Issue by dduan** [swiftlang/swift-package-manager#8140]
### Description

tl;dr `swift install` is a command that builds the default execuatble target in release mode, and copy the result somewhere.

The *somewhere* should default to a fixed location that the user can access (for example, `~/swift/bin`). The user can access it directly, or add it to their `PATH`. This location can be overriden by a envvar, for example, `$SWIFTBIN`.

Why is this a great idea? Sometimes you don't want to distribute your CLI applications via a public channel such as Homebrew. Maybe it's not meant for others to use, maybe you want to control what version to use by building from source.

Imagine a in-house tool whose source is part of a repository. The tool is meant to be version locked with whatever the repo's git revision is.

```
monorepo
├── ...
├── tools
│   └── great-swift-tool
│       └── Package.swift
└── tmp # git ignored
    └── bin
```

How does one build `great-swift-tool` and put it under `./tmp/bin` today? Easy:

```
swift build --package-path tools/great-swift-tool -c release
cp tools/great-swift-tool/.build/release/great-swift-tool-executable tmp/bin
```

... put that in a shell script or make target and forget it, right?

What if there are 10 of these tools? What if you need to move their source code around in the repo for some reason?

IMO this is unnecessarily complicated in 2 ways:

1. the user has to know that the build result is in `.build/release`
2. the user has to write out the executable's name

With the proposed solution, this could be simplified as 1 command:

```
SWIFTBIN=tmp/bin swift install --package-path tools/great-swift-tool
```

I think that'd be pretty cool.

**Comment by kkebo** [comment]
There is already `experimental-`-prefixed `swift package install`.

https://github.com/swiftlang/swift-package-manager/pull/6768

**Comment by unnamedd** [comment]
A good inspiration can be Cargo (from Rust). 
There, we have the same possibilities that were brought here.

Here is the help from `cargo install`:

```shell
➜ cargo install --help
Install a Rust binary

Usage: cargo install [OPTIONS] [CRATE[@<VER>]]...

Arguments:
  [CRATE[@<VER>]]...  Select the package from the given source

Options:
      --version <VERSION>     Specify a version to install
      --index <INDEX>         Registry index to install from
      --registry <REGISTRY>   Registry to use
      --git <URL>             Git URL to install the specified crate from
      --branch <BRANCH>       Branch to use when installing from git
      --tag <TAG>             Tag to use when installing from git
      --rev <SHA>             Specific commit to use when installing from git
      --path <PATH>           Filesystem path to local crate to install from
      --root <DIR>            Directory to install packages into
  -f, --force                 Force overwriting existing crates or binaries
      --no-track              Do not save tracking information
      --list                  List all installed packages and their versions
      --message-format <FMT>  Error format
      --debug                 Build in debug mode (with the 'dev' profile) instead of release mode
  -v, --verbose...            Use verbose output (-vv very verbose/build.rs output)
  -q, --quiet                 Do not print cargo log messages
      --color <WHEN>          Coloring: auto, always, never
      --config <KEY=VALUE>    Override a configuration value
  -Z <FLAG>                   Unstable (nightly-only) flags to Cargo, see 'cargo -Z help' for details
  -h, --help                  Print help
```

**Comment by czechboy0** [comment]
I'd welcome the even more streamlined version from Go:

```
go install github.com/myorg/mytool@latest
```

I usually _don't_ want to clone the sources first, and then build separately. So ideally, I'd like to have:

```
swift install github.com/myorg/mytool@latest
```

**Comment by dschaefer2** [comment]
I'm thinking bigger, like `cmake --install`, which populates a whole binary tree.

It's funny since the original SwiftPM proposal 10 years ago had the capability you're talking about but they decided they didn't want to overlap with git. But hard to argue the developer experience of what `go` gives you here.

The good news is that we're trying to change the mindset of SwiftPM away from being library centric to being a more general build system. This definitely fits my vision for it.
